In [2]:
# importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import label_binarize

# importing the dataset
data = pd.read_csv('./Iris.csv', index_col=0)
data.head()

#data["isSetosa"] = data["Species"]
#data.head()

species_dictionary = {'Iris-setosa': 'setosa', 'Iris-versicolor': 'not-setosa', 'Iris-virginica': 'not-setosa'}
data = data.replace({"Species": species_dictionary})
data.sample(5)

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm     Species
Id                                                                       
96             5.7           3.0            4.2           1.2  not-setosa
112            6.4           2.7            5.3           1.9  not-setosa
46             4.8           3.0            1.4           0.3      setosa
82             5.5           2.4            3.7           1.0  not-setosa
138            6.4           3.1            5.5           1.8  not-setosa

In [3]:
#binarize
data["Species"] = (data["Species"] != "setosa").astype(int)

#split into train and test
#y = binary for isSetosa
X = data.iloc[:,[0,1,2,3]].values
y = data.iloc[:,4].values

#from sklearn.preprocessing import LabelBinarizer
#lb = LabelBinarizer()
#y = lb.fit_transform(y)


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)


In [4]:
data.sample(5)

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
Id                                                                    
41             5.0           3.5            1.3           0.3        0
58             4.9           2.4            3.3           1.0        1
139            6.0           3.0            4.8           1.8        1
109            6.7           2.5            5.8           1.8        1
40             5.1           3.4            1.5           0.2        0

In [5]:
# fit a model
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

# make predictions on test data
y_pred = log_reg.predict(X_test).reshape(-1,1)

In [6]:
# use score method to get accuracy of model
score = log_reg.score(X_test, y_test)

print('Accuracy: {}'.format(score))

Accuracy: 1.0


Accuracy of 1.0 says that it is a 100% accuracy.

In [7]:
iris = load_iris()
#data['Species'] = iris.target
data.sample(6)

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
Id                                                                    
102            5.8           2.7            5.1           1.9        1
126            7.2           3.2            6.0           1.8        1
139            6.0           3.0            4.8           1.8        1
76             6.6           3.0            4.4           1.4        1
124            6.3           2.7            4.9           1.8        1
117            6.5           3.0            5.5           1.8        1

In [8]:
from sklearn.metrics import confusion_matrix

classes = list(data['Species'].unique())
confusion_matrix = confusion_matrix(y_test, y_pred)
#cm_df = pd.DataFrame(confusion_matrix, columns=classes, index=classes)
#cm_df

In [9]:
classes

[0, 1]

In [10]:
confusion_matrix = pd.DataFrame(confusion_matrix, columns=classes)
confusion_matrix

0   1
0  11   0
1   0  19

In [11]:
data.shape

(150, 5)

comment prediction on higher precision, higher recall or both similar
precision = proportion of predictions of positive class that is correct (of all predictions that the iris is a setosa, what proportion was correct?
TP / (TP + FP)
recall = how many instances of a class the model was able to recognise. Of all actual setosas, what proportion of them was spotted by the model?
TP / (TP + FN)

True Label (vertical) Predicted label (horizontal)

11 setosa were predicted as setosas, 0 setosas were predicted as non-setosas
0 non-setosas were predicted as setosas, 19 non-setosas were predicted as non-setosas

there were 150 entries in the data
I predict the precision for this model will be higher than the recall, as a relatively small number of setosas was identified, so it is likely that most of them are correct, however the rest of the setosas in the dataset was not identified. Therefore I suspect a low number of false positives, as the model seems fairly strict, therefore TP / (TP + FP) will yield a higher number than TP / (TP + FN)



In [12]:
#calculate accuracy, precision and recall with your own code
#Accuracy = (TP + TN) / (TP + FN + TN + FP)
#Precision = TP / (TP + FP)
#Recall = TP / (TP + FN)

#Binary Confusion Matrix:
#| TP | FP |       | 11 | 0 |
#| FN | TN |       | 0 | 19 |

confusion_matrix

0   1
0  11   0
1   0  19

In [23]:
TP = confusion_matrix.iloc[0,0]
FP = confusion_matrix.iloc[0,1]
FN = confusion_matrix.iloc[1,0]
TN = confusion_matrix.iloc[1,1]

In [27]:
Accuracy = (TP + TN) / (TP + FN + TN + FP)
Accuracy

1.0

In [28]:
Precision = TP / (TP + FP)
Precision

1.0

In [29]:
Recall = TP / (TP + FN)
Recall

1.0

Turns out the Precision and Recall are the same for my model because my values for FP and FN are 0, so it makes sense that the model appears to be 100% correct with its classifications when all of its classifications are True (True positives, correctly identifying Setosas, and True Negatives, correctly identifying non-setosas)
This might indicate that the model's classification is too strict and does not leave room for False Positive or False Negative predictions, which means that other True negatives and True positives were also missed out as they did not meet the strict classification requirements.